stub association, tq and kf update.

tschuh opened the pull request:
This PR moves the StubAssociation class from SimTracker/... to SimDataFormats/... where it belongs. The algorithm to associate Stubs with TPs has moved to an ESProduct called StubAssociator in L1Trigger/TrackTrigger using the previous mentioned  StubAssociation class. There are now 3 Association maps to measure fake rate, duplicate rate and efficiency. fake rate definition has changed (treating a particle after scattering not as a new particle) to reduce false fakes. Analyzer has been updated accordingly.

An EDProduer Cleaner has been added in SimTracker/TrackTriggerAssociation, it produces a collection of TrackingVertices & TrackingParticles which I call clean (only contains TPs which either are associated with at least on cluster or are a parent/child of such a TP). Additionally it provides a TTClusterAssociationMap pointing to this new collection.

Some boiler plate changes in EDProducers (removing unnecessary template parameter)

DataFormats contain now TrackQulaity output data formats. TQ and TFP code has been updated accordingly.

L1FPGATrackProducer calls directly TQ simulation if wanted.

TrackerTFP TrackQuality simulation has been removed (since it is now in L1FPGATrackProducer)

TrackFindingTracklet TrackQuality has been created in preparation of syncing with fw.

UPDATE 20/08/2025: TrackFindingTracklet analyzer factorized and moved to TrackTrigger.

UPDATE 21/08/2025: added a third Association map for duplicate rate calculation.

UPDATE 19/09/2025: kf synced with f/w

UPDATE 26/09/2025: TM bug fix of stub tilted bit, impacting stub uncertainty calculation in DR

UPDATE 25/11/2025: TFP removes all gaps

for Ian https://indico.cern.ch/event/1587186/

tschuh commented:
@tomalin could you please fix the CI or show me where I can fix it?

tomalin commented:
This PR has changed 93 files. Is the description of it complete? e.g. I see that all the analyzers have been moved out of TrackFindingTracklet. And it's not explained why a 3rd association map is needed for the duplicate rate calculation.

amastron23 commented:
> TrackFindingTracklet TrackQuality has been created in preparation of syncing with fw.

@tschuh Is this done in this branch? and does that come with firmware changes too?

tomalin commented:
Please check that the comments at the top of each class explain what it does. e.g. ProducerFakeTM and ProducerFakeDR currently have identical comments, but I assume they do different things?

tomalin commented:
ProducerFakeTM.cc creates an object of type TrackTM, which only takes as input the 3 helix params inv2R, phiT, zT. It is used by HYBRID_NEWKF_DISPLACED_KILL. Does this mean that the tracklet estimates of z0 and d0 are ignored? Similarly ProducerFakeDR.cc creates an object of type TrackDR, which takes the same 3 helix params. It is used by HYBRID_NEWKF_DISPLACED_MERGE, so are tracklet estimates of z0 and d0 ignored here too?

tschuh commented:
> Please check that the comments at the top of each class explain what it does. e.g. ProducerFakeTM and ProducerFakeDR currently have identical comments, but I assume they do different things?

Well spotted, fixed.

tschuh commented:
> ProducerFakeTM.cc creates an object of type TrackTM, which only takes as input the 3 helix params inv2R, phiT, zT. It is used by HYBRID_NEWKF_DISPLACED_KILL. Does this mean that the tracklet estimates of z0 and d0 are ignored? Similarly ProducerFakeDR.cc creates an object of type TrackDR, which takes the same 3 helix params. It is used by HYBRID_NEWKF_DISPLACED_MERGE, so are tracklet estimates of z0 and d0 ignored here too?

z0 is used to calculate zT, so it is not ignored, d0 gets ignored.

tschuh commented:
> This PR has changed 93 files. Is the description of it complete? e.g. I see that all the analyzers have been moved out of TrackFindingTracklet. And it's not explained why a 3rd association map is needed for the duplicate rate calculation.

I updated the description.

I believe this PR fixes bugs in the KF (5 param fit not finding tracks with d0 > 1 cm, 4 param fit having poor z0 resolution). It fixes the TQ which was not working. Please ensure the description describes important changes that affect people.

tschuh commented:
> > TrackFindingTracklet TrackQuality has been created in preparation of syncing with fw.
> 
> @tschuh Is this done in this branch? and does that come with firmware changes too?

now it is, f/w should be unchanged.

tomalin commented:
Please add link to your talk https://indico.cern.ch/event/1587186/ to description of this PR.

tomalin commented:
The current definitions of efficiency, fake & duplicate rates are the agreed ones within the L1 track community, and give consistent results between your analyzers and the original associators and L1TrackNTuplePlot.C. There was not support in the BES SW meeting for adopting your definitions as the default. Therefore, please ensure that by default, the default results for efficiency, fake & duplicate rates are preserved. 
In addition, it is silly having two sets of associators. The one you developed has the advantage that it easily allows associating L1 tracks and TP part way through the chain, whereas this is a bit more painful with the original one, since the EDProducts of those associators do not allow more than one incorrect stub. We could consider making yours the default, and removing the other one at some point in the future. But this would require that yours runs by default on the standard TP collection, since that is the only one that is persistent.

tomalin commented on L1Trigger/TrackFindingTracklet/plugins/ProducerTQ.cc:
```diff
@@ -0,0 +1,139 @@
+    int shift;
+    double base;
+    double range;
+    // m02
```
Please say physically what the m, invV and the chi variables here are.

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc:
```diff
@@ -832,6 +838,62 @@ void L1FPGATrackProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSe
 
 }  /// End of produce()
 
+// Controls the conversion between TTTrack features and ML model training features
```
Is featureTransform() used only by the bit-accurate TQ emulation or also by the TQ simulation? 
This is rather long code, so I'd prefer to see it moved into a separate class/function, rather than extending the already very long L1FPGATrackProducer.

tschuh replied:
this is only called by simulation. I find it small compared to all other interfaces in this Producer. But I can put it in a separate header file when you really fancy it.

tomalin replied:
Yes, please move featureTransform() to separate file, and add comment at top explaining what it does, and mention there that this is only used by the TQ simulation, not the TQ emulation.

tschuh replied:
done

tomalin commented:
Most of the EDAnalyzers in TrackFindingTracklet have been deleted. Please explain in the PR description what is going on.

tomalin commented on L1Trigger/TrackFindingTracklet/test/BuildFile.xml:
```diff
@@ -1,7 +1,7 @@
   <flags   EDM_PLUGIN="1"/>
   <flags   SKIP_FILES="fpga.cc"/>
-  <flags   CXXFLAGS="-DUSEHYBRID"/>
+  <flags   CXXFLAGS="-DUSEHYBRID -lTMVA"/>
```
What's the -lTMVA flag for here?

tschuh replied:
ah, I played with bdt training in cmssw, will remove it

tomalin commented on SimTracker/TrackTriggerAssociation/python/Cleaner_cff.py:
```diff
@@ -0,0 +1,5 @@
+import FWCore.ParameterSet.Config as cms
+
```
All python files should have a comment at the top explaining what they are for.

tschuh replied:
obsolete

tomalin commented on SimTracker/TrackTriggerAssociation/plugins/Cleaner.cc:
```diff
@@ -0,0 +1,178 @@
+namespace tt {
+
+  /*! \class  tt::Cleaner
+   *  \brief  creates TrackingParticleCollection with at least one direct or indirect
```
Suggest modifying comment: Change to "Creates filtered TrackingParticleCollection, where only TP with at least ... are retained."

tschuh replied:
obsolete

tomalin commented on SimTracker/TrackTriggerAssociation/plugins/Cleaner.cc:
```diff
@@ -0,0 +1,178 @@
+   *          (via parent or child) associated cluster.
+   *          creates TrackingVertexCollection with all vertices (with corrected
+   *          parent TP connection) for those TPs.
+   *          creates TTClusterAssociationMap with those TPs
```
Suggest changing comment to "creates reduced copy of the TTClusterAssociation map, which contains only those TPs".
Also, explain purpose of this code. e.g. Save CPU by reducing size of collections. Avoids labelling L1 tracks as fake, if they are formed from stubs created by a TP decaying to another TP.

tschuh replied:
obsolete

tomalin commented on L1Trigger/TrackTrigger/interface/Associator.h:
```diff
@@ -0,0 +1,84 @@
+namespace tt {
+
+  /*! \class  tt::Associator
+   *  \brief  Class to associate TrackingParticles with TTStubs and vice versa.
```
Explain how this differs from the other TP to stub associator.

tomalin replied:
Why has this class been moved? It should be in the same directory as the other associators, and its original name was better, as it explained what it was associating.

tschuh replied:
the other TP to stub associator is an EDProducer, this is an ESProducer. It is in L1Trigger/TrackTrigger since it depends on tt::Setup.

tomalin commented on L1Trigger/TrackTrigger/plugins/ProducerAssociator.cc:
```diff
@@ -0,0 +1,46 @@
+namespace tt {
+
+  /*! \class  tt::ProducerAssociator
+   *  \brief  provides class to associate TrackingParticles with TTStubs and vice versa.
```
Why has this class been moved? It should be in the same directory as the other associators. Also, in comment, explain how differs from other TP to stub associatior.

tschuh replied:
wrong, associations belong in SimDataFormats/Associations. Also comment already explains the difference ...

tomalin replied:
The Products of the association belong in SimDataFormats/Associations, so you have correctly put class StubAssociation there. But the Tracker code that calculates the associations has traditionally always been in SimTracker/TrackTriggerAssociation/ . Isn't putting your associator calculator in a different directory to all the others making it harder for people to realize that there is more than one and to compare them? Is the only reason  you've moved it that it depends on Setup, and you're concerned that SimTracker/TrackTriggerAssociation may not be allowed to depend on L1Trigger/TrackTrigger? 
Also what is the advantage of using an ESProducer rather than an EDProducer for the association?

tschuh replied:
The Class which produces StubAssociation is called StubAssociator and is placed in SimTracker/TrackTriggerAssociation/. Obliviously your complaint doesn't make any sense.

tomalin commented on L1Trigger/TrackTrigger/python/Associator_cfi.py:
```diff
@@ -0,0 +1,12 @@
+import FWCore.ParameterSet.Config as cms
+
```
All python files must have a comment explaining what they are.

tschuh replied:
done

tomalin replied:
Still no comment added

tschuh replied:
upsi, done now

tomalin commented on L1Trigger/TrackTrigger/test/AnalyzerMC.cc:
```diff
@@ -0,0 +1,166 @@
+
+namespace tt {
+
+  class AnalyzerMC : public edm::one::EDAnalyzer<edm::one::WatchRuns, edm::one::SharedResources> {
```
All classes need a comment at the top explaining what they do.

tschuh replied:
done

tomalin commented on L1Trigger/TrackTrigger/test/AnalyzerMC_cfg.py:
```diff
@@ -0,0 +1,48 @@
+import FWCore.ParameterSet.Config as cms
+
```
All jobs need a comment at the top explaining what they do. As we have several jobs analysing the tracking, it also needs to explain specifically what this one does.

tschuh replied:
obsolete

tomalin commented on SimDataFormats/Associations/interface/StubAssociation.h:
```diff
@@ -0,0 +1,59 @@
+
+  /*! \class  tt::StubAssociation
+   *  \brief  Class to store maps to associate TrackingParticles with TTStubs and vice versa.
+   *          It may associate multiple TPs with a TTStub and can therefore be used to associate
```
Add coment explaining difference to standard stub associator.

tschuh replied:
it is already there: It may associate multiple TPs with a TTStub and can therefore be used to associate TTTracks with TrackingParticles.

tomalin commented:
I'd still like to see a decent README for your code, explaining how to use things like the formatter. It's still the case that very few people have a deep enough understanding of it to be able to make significant modifications or improvements to it, or even to help maintain it. A README may help.

tomalin commented:
Needs a rebase to fix git CI.

tomalin commented:
How did you rebase this? My impression is that this hasn't been done correctly. Changes made by other people have been lost. (Or perhaps you haven't rebased it yet?)

tomalin commented:
I just checked out tschuh branch and ran L1TrackNtupleMaker_cfg.py for L1TRKALGO = HYBRID_DISPLACED_NEWKF_KILL. It failed with the following error message, (whilst it runs OK in the default branch):

An exception of category 'ScheduleExecutionFailure' occurred while
   [0] Calling beginJob
Exception Message:
Unrunnable schedule
Paths are non consistent
  module 'StubAssociator' depends on 'Cleaner' which appears on paths
  dtc 
but is missing from
  TTTracksEmulationWithTruth

amastron23 commented:
@tschuh 

I am getting this error when running demonstrator_cfg.py

<img width="643" height="157" alt="image" src="https://github.com/user-attachments/assets/4fc59be3-ed9e-4781-8bbe-e721b8b222b9" />

Not a showstopper though from my pov. The text files do get generated and they look correct to my eye.

meisonlikesicecream commented:
> @tschuh
> 
> I am getting this error when running demonstrator_cfg.py
> 
> <img alt="image" width="643" height="157" src="https://private-user-images.githubusercontent.com/145787018/507750144-4fc59be3-ed9e-4781-8bbe-e721b8b222b9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE5MjA0MjMsIm5iZiI6MTc2MTkyMDEyMywicGF0aCI6Ii8xNDU3ODcwMTgvNTA3NzUwMTQ0LTRmYzU5YmUzLWVkOWUtNDc4MS04YmJlLWU3MjFiOGIyMjJiOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAzMVQxNDE1MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNWEwZGVmNTE3ZTRhYTc3OGQ2ODUwODZhYmY3YzAyMjMyNGJkYzMzM2Y4YjQ0MDI0YjFmYjZjMDg4NDk2ODY1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0e17uHufaqmGEKmXK6D2sQudu1HdZX7X0Zy68A2jKoA">
> Not a showstopper though from my pov. The text files do get generated and they look correct to my eye.

@tschuh, I explained this error (probably because modelsim/questasim was not sourced and couldn't run vsim) to @amastron23. Not just me being confused ;)

tschuh committed:
squash before rebase

tschuh committed:
rebase fixes

tschuh committed:
fixes to run.

tschuh committed:
default analyzer change for fake and duplicate rate.

tschuh commented:
> I just checked out tschuh branch and ran L1TrackNtupleMaker_cfg.py for L1TRKALGO = HYBRID_DISPLACED_NEWKF_KILL. It failed with the following error message, (whilst it runs OK in the default branch):
> 
> An exception of category 'ScheduleExecutionFailure' occurred while [0] Calling beginJob Exception Message: Unrunnable schedule Paths are non consistent module 'StubAssociator' depends on 'Cleaner' which appears on paths dtc but is missing from TTTracksEmulationWithTruth

fixed

tschuh commented:
> How did you rebase this? My impression is that this hasn't been done correctly. Changes made by other people have been lost. (Or perhaps you haven't rebased it yet?)

rebased yesterday

tschuh committed:
Ian 0

tschuh committed:
clean up & little fixes.

tschuh committed:
ProducerFakeTM fix

tschuh committed:
code-format

tomalin commented:
I just ran L1TrackNtupleMaker_cfg.py with option HYBRID_DISPLACED_NEWKF_KILL on the displaced muon MC. According to L1TrackNtuplePlot.C, it does now have good efficiency for d0 > 1cm, which is excellent progress.
However, the printed report produced by your code is buggy. e.g. The DTC SUMMARY contains the lines:
"max tracking efficiency = -nan +- -nan", as do all the other L1 track SUMMARYs (which also show "current tracking efficiency = -nan +- -nan".

tschuh commented:
well you told the StubAssociator to don't take with TPs with large d0 or vertex r into account. The bug is that you did not overwrote that. Which cuts shall we use for displaced tracking?

tschuh committed:
FakeTM improved and displaced TP cuts loosened.

tschuh committed:
code format.

tomalin commented on L1Trigger/TrackFindingTracklet/test/AnalyzerTB.cc:
```diff
@@ -0,0 +1,341 @@
+    consume(streamsTrack, streamsStub, tracks, stubs);
+    // analyze in tracks and stubs
+    for (const Track& track : tracks) {
+      std::vector<TTStubRef> stubs;
```
Here one gets stubs from function consume, and then immediately defines a new variable stubs of different type, but with the same name. This is confusing. Please use a different variable name.

tschuh replied:
done

tomalin commented on L1Trigger/TrackFindingTracklet/test/AnalyzerTB.cc:
```diff
@@ -0,0 +1,341 @@
+    void endJob() override {}
+
+  private:
+    struct Stub {
```
We already have locks of classes called Stub, such as the one in DataFormats.h, the one in Stub.h, and now we have this one struct too. This is confusing, as the reader doesn't know (without effort) which one is being used. Can we have a slightly different name? Same may also apply to Track. And I guess this comment may apply to other EDAnalyzers too.

tschuh replied:
done

tschuh committed:
Ian 1

tschuh committed:
code-format

tschuh committed:
tfp gap removal.

tschuh committed:
ProducerFakeTM bug fix

tschuh committed:
added comments to AnalyzerTB and displaced KF fix

tschuh committed:
displaced improvement

tomalin committed:
Optimize CPU for displaced tracking (#342)

* Optimize CPU for displaced tracking

* code format

tomalin commented:
> > Please check that the comments at the top of each class explain what it does. e.g. ProducerFakeTM and ProducerFakeDR currently have identical comments, but I assume they do different things?
> 
> Well spotted, fixed.

You haven't fixed it.

tomalin commented:
KalmanFilter.h now has options to run 4 or 5 parameters fits plus the old KF simulation. Please state this in the header. Also make clear that the 4-param fit is bit-accurate and includes a KF seeding stage, whereas the other two do not. Explain how to switch between them.

tschuh commented:
> > > Please check that the comments at the top of each class explain what it does. e.g. ProducerFakeTM and ProducerFakeDR currently have identical comments, but I assume they do different things?
> > 
> > 
> > Well spotted, fixed.
> 
> You haven't fixed it.

can't confirm, I changed KF to DR in ProducerFakeTM.

tomalin commented:
> > ProducerFakeTM.cc creates an object of type TrackTM, which only takes as input the 3 helix params inv2R, phiT, zT. It is used by HYBRID_NEWKF_DISPLACED_KILL. Does this mean that the tracklet estimates of z0 and d0 are ignored? Similarly ProducerFakeDR.cc creates an object of type TrackDR, which takes the same 3 helix params. It is used by HYBRID_NEWKF_DISPLACED_MERGE, so are tracklet estimates of z0 and d0 ignored here too?
> 
> z0 is used to calculate zT, so it is not ignored, d0 gets ignored.

I still don't get this. This line https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/plugins/ProducerFakeTM.cc#L124 seems to use only 3 helix parameters to define the initial track parameters that are fed into the KF fit. Surely 4 helix parameters are needed? All the tracklet hit residuals will have been calculated with respect to the state defined by 4 helix params, so they can't be interpreted if only 3 helix params are retained.

tomalin commented:
> Please add link to your talk https://indico.cern.ch/event/1587186/ to description of this PR.

Still to do.

tschuh commented:
> > > ProducerFakeTM.cc creates an object of type TrackTM, which only takes as input the 3 helix params inv2R, phiT, zT. It is used by HYBRID_NEWKF_DISPLACED_KILL. Does this mean that the tracklet estimates of z0 and d0 are ignored? Similarly ProducerFakeDR.cc creates an object of type TrackDR, which takes the same 3 helix params. It is used by HYBRID_NEWKF_DISPLACED_MERGE, so are tracklet estimates of z0 and d0 ignored here too?
> > 
> > 
> > z0 is used to calculate zT, so it is not ignored, d0 gets ignored.
> 
> I still don't get this. This line https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/plugins/ProducerFakeTM.cc#L124 seems to use only 3 helix parameters to define the initial track parameters that are fed into the KF fit. Surely 4 helix parameters are needed? All the tracklet hit residuals will have been calculated with respect to the state defined by 4 helix params, so they can't be interpreted if only 3 helix params are retained.

Nope, we work with 3 and assume z0 = 0 and shift the residuals accordingly.

tomalin committed:
Possibly fix memory leak crash

tschuh committed:
Ian 2

tschuh commented:
> KalmanFilter.h now has options to run 4 or 5 parameters fits plus the old KF simulation. Please state this in the header. Also make clear that the 4-param fit is bit-accurate and includes a KF seeding stage, whereas the other two do not. Explain how to switch between them.

done

tschuh commented:
> > Please add link to your talk https://indico.cern.ch/event/1587186/ to description of this PR.
> 
> Still to do.

done

tschuh committed:
code format of ian commit

tschuh committed:
moved featureTransform into seperate file.

tomalin commented:
> I'd still like to see a decent README for your code, explaining how to use things like the formatter. It's still the case that very few people have a deep enough understanding of it to be able to make significant modifications or improvements to it, or even to help maintain it. A README may help.

We have a README for the new Hough Transform algo /L1Trigger/TrackerTFP/README.md and this does describe briefly some of the classes such as DataFormats used also by HYBRID. But it's unlikely that the HYBRID people will look for the documentation there, unless one of the other README that are used for HYBRID explicitely tells them to look.

tomalin commented:
> > > > ProducerFakeTM.cc creates an object of type TrackTM, which only takes as input the 3 helix params inv2R, phiT, zT. It is used by HYBRID_NEWKF_DISPLACED_KILL. Does this mean that the tracklet estimates of z0 and d0 are ignored? Similarly ProducerFakeDR.cc creates an object of type TrackDR, which takes the same 3 helix params. It is used by HYBRID_NEWKF_DISPLACED_MERGE, so are tracklet estimates of z0 and d0 ignored here too?
> > > 
> > > 
> > > z0 is used to calculate zT, so it is not ignored, d0 gets ignored.
> > 
> > 
> > I still don't get this. This line https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/plugins/ProducerFakeTM.cc#L124 seems to use only 3 helix parameters to define the initial track parameters that are fed into the KF fit. Surely 4 helix parameters are needed? All the tracklet hit residuals will have been calculated with respect to the state defined by 4 helix params, so they can't be interpreted if only 3 helix params are retained.
> 
> Nope, we work with 3 and assume z0 = 0 and shift the residuals accordingly.

OK, then I'd like a comment in the code explicitely stating that, since throwing away the tracklet z0 info does sound a strange thing to do, so people are unlikely to guess that the plan is to do that.

tschuh committed:
ian 3

tschuh commented:
> > I'd still like to see a decent README for your code, explaining how to use things like the formatter. It's still the case that very few people have a deep enough understanding of it to be able to make significant modifications or improvements to it, or even to help maintain it. A README may help.
> 
> We have a README for the new Hough Transform algo /L1Trigger/TrackerTFP/README.md and this does describe briefly some of the classes such as DataFormats used also by HYBRID. But it's unlikely that the HYBRID people will look for the documentation there, unless one of the other README that are used for HYBRID explicitely tells them to look.

I updated the README file

tschuh commented:
> > > > > ProducerFakeTM.cc creates an object of type TrackTM, which only takes as input the 3 helix params inv2R, phiT, zT. It is used by HYBRID_NEWKF_DISPLACED_KILL. Does this mean that the tracklet estimates of z0 and d0 are ignored? Similarly ProducerFakeDR.cc creates an object of type TrackDR, which takes the same 3 helix params. It is used by HYBRID_NEWKF_DISPLACED_MERGE, so are tracklet estimates of z0 and d0 ignored here too?
> > > > 
> > > > 
> > > > z0 is used to calculate zT, so it is not ignored, d0 gets ignored.
> > > 
> > > 
> > > I still don't get this. This line https://github.com/cms-L1TK/cmssw/blob/tschuh/L1Trigger/TrackFindingTracklet/plugins/ProducerFakeTM.cc#L124 seems to use only 3 helix parameters to define the initial track parameters that are fed into the KF fit. Surely 4 helix parameters are needed? All the tracklet hit residuals will have been calculated with respect to the state defined by 4 helix params, so they can't be interpreted if only 3 helix params are retained.
> > 
> > 
> > Nope, we work with 3 and assume z0 = 0 and shift the residuals accordingly.
> 
> OK, then I'd like a comment in the code explicitely stating that, since throwing away the tracklet z0 info does sound a strange thing to do, so people are unlikely to guess that the plan is to do that.

done

tschuh committed:
TFP output scrambled according to specification.

tomalin commented:
This looks good to go, as soon as it passes git CI.

tomalin reviewed (approved):
This PR is great progress and looks good. Approved.
